### DISCUSSION TITLE
Change the page creation in the PageNavigationService

### DISCUSSION THREAD

**Discussion by Grzesik** [PrismLibrary/Prism#2247] [Xamarin.Forms]
Hi,
in one of my Xamarin.Forms projects I create the Xaml pages dynamically (from a text file). To use Prism for the project  I inherited a class from the class PageNavigationService and overrode the function 

`protected override Page CreatePage(string segmentName)`

In the Prism Version 7 the function was quite easy to override:

```
protected virtual Page CreatePage(string segmentName)
        {
            try
            {
                return _container.Resolve<object>(segmentName) as Page;
            }
            catch (Exception ex)
            {
                if (((IContainerRegistry)_container).IsRegistered<object>(segmentName))
                    throw new NavigationException(NavigationException.ErrorCreatingPage, _page, ex);

                throw new NavigationException(NavigationException.NoPageIsRegistered, _page, ex);
            }
        }

```

instead of the container.Resolve<object>(segmentName) as Page; I created the page dynamically, like this new ContentPage().LoadFromXaml(xaml);

In the Prism version 8 the function is more complicated:

```
protected virtual Page CreatePage(string segmentName)
        {
            try
            {
                _container.CreateScope();
                var page = (Page)_container.Resolve<object>(segmentName);

                if (page is null)
                    throw new NullReferenceException($"The resolved type for {segmentName} was null. You may be attempting to navigate to a Non-Page type");

                return SetNavigationServiceForPage(page);
            }
            catch (Exception ex)
            {
                if (((IContainerRegistry)_container).IsRegistered<object>(segmentName))
                    throw new NavigationException(NavigationException.ErrorCreatingPage, _page, ex);

                throw new NavigationException(NavigationException.NoPageIsRegistered, _page, ex);
            }
        }
```

To override the function, I needed to implement the function SetNavigationServiceForPage, because it is private. Unfortunately in the function is the Xaml.Navigation.NavigationServiceProperty internal and implemented as a static class. I needed also to copy the Navigation class.

```
private Page SetNavigationServiceForPage(Page page)
        {
            // Someone explicitly set Autowire ViewModel
            if (page.GetValue(Xaml.Navigation.NavigationServiceProperty) != null)
                return page;

            // This will wireup the Navigation Service in case you have something injected that
            // actually required the Nav Service
            var childNavService = _container.Resolve<INavigationService>();
            if (childNavService is IPageAware pa)
                pa.Page = page;

            page.SetValue(Xaml.Navigation.NavigationServiceProperty, childNavService);
            return page;
        }
```

I suggest to change the implementation of the CreatePage function. The easiest way is to change the function SetNavigationServiceForPage to protected. The other possibility is to create a virtual function for a part of the the page creation:

```
protected Page CreateRawPage()
{
   _container.CreateScope();
   var page = (Page)_container.Resolve<object>(segmentName);
   return page;
}
```

and use it in the _CreatPage_ function.

Thanks 
Marian

**Comment by dansiegel** [comment]
I'm not going to do a CreateRawPage, but we can make the private method protected that way you can access it in your override.

**Reply by Grzesik** [reply]
I think it'd be than fine for me to implement my own dynamic page creation. Thanks :-) Merry Christmas
